File: src/main/scala/org/hejwo/scala/algorithms/sorting/atest/own/inheritance/java/Customer.java
package org.hejwo.scala.algorithms.sorting.atest.own.inheritance.java;

import com.google.common.collect.Lists;
import org.apache.commons.lang3.tuple.Pair;

import java.util.List;

public class Customer implements Comparable<Customer> {

    private final String name;
    private final String address;
    private final List<Pair<String, Double>> basket = Lists.newArrayList();

    public Customer(String name, String address) {
        this.name = name;
        this.address = address;
    }

    public void add(String item, Double price) {
        basket.add(Pair.of(item, price));
    }

    public Double total() {
        return basket.stream().mapToDouble(Pair::getRight).sum();
    }

    @Override
    public int compareTo(Customer o) {
        return this.name.compareTo(o.name);
    }

    @Override
    public String toString() {
        return "Customer{" +
                "name='" + name + '\'' +
                '}';
    }
}

File: src/main/scala/org/hejwo/scala/algorithms/sorting/atest/own/inheritance/scala/Customer.scala
package org.hejwo.scala.algorithms.sorting.atest.own.inheritance.scala

class Customer(val name: String, val address: String) extends Ordered[Customer] {

  private final var basket: List[(String, Double)] = List()

  def add(name: String, price: Double) = {
    basket = (name, price) :: basket
  }

  def total: Double = basket.map(_._2).sum

  override def compare(that: Customer) = name.compareTo(that.name)

  override def toString = s"Customer($name)"
}

class DiscountedCustomer(name: String, address: String) extends Customer(name, address) {
  override def total = super.total * 0.90
}